### DISCUSSION TITLE
@Arg undefined in @FieldResolver

### DISCUSSION THREAD

**Discussion by cnps** [MichalLytek/type-graphql#1158] [Q&A] [answered]
Following various examples, it seems @Arg from parent query, is not getting passed to @FieldResolver.

If i'm trying to access @Arg in the @FieldResolver, it won't run the query and returns an error: _"[GraphQL error]: Message: Field "distanceToProvider" argument "location" of type "[Int!]!" is required, but it was not provided."_

It works fine in the parent query, if I remove @Arg from @FieldResolver

```
@Resolver(Provider)
export class ProviderResolver {
  @Query((returns) => [Provider])
  async providers(
    @Arg('location', () => [Int]) location: number[]
  ): Promise<Provider[]> {
    const providers: Provider[] = await ProviderModel.find();
    
    // @Arg value available here

    return providers;
  }

  @FieldResolver(() => Int)
  async distanceToProvider(
    @Root() provider: DocumentType<Provider>,
    @Arg('location', () => [Int]) location: number[]
  ): Promise<number> {
        // @Arg value is undefined here
        // calculate distance to provider, based on users current location
    return location;
  }
}
```

Here is the generated model
```
"""Model for the provider"""
type Provider {
  _id: ID!
  aws_id: String!
  createdAt: DateTime!
  deletedAt: DateTime!
  distanceToProvider(location: [Int!]!): Float!
  email: String!
  img: String
  locations: [Location!]!
  name: String!
  openingHours: [OpeningHour!]!
  phonenumber: String
  products: [ProviderProduct!]!
  profession: Profession!
  pusers: [PUser!]!
  rating: Rating
  reviewCount: Float!
  reviews: [Review!]
  settings: ProviderSettings!
  updatedAt: DateTime!
  website: String
}
```

Am I doing it correct? Or is there an alternative to get @Arg/@Args in the fieldresolver?

**Comment by MichalLytek** [comment, marked as answer]
> Following various examples, it seems @arg from parent query, is not getting passed to @FieldResolver.

Because that's how GraphQL was designed. Fields have own arguments and should never rely on arguments of the parent query because they are independent and field resolver can be executed as a part of other query.

**Reply by cnps** [reply]
Thank for the quick reply! It seems I have misunderstood what a @FieldResolver really is. I guess I only need a FieldResolver, if I need to fetch referenced relational data? I guess I was confused by the "averageRating" example in the docs

It seems something like this works. Is it good practice to mutate db response before returning it in the resolver?

```
  @Query((returns) => [Provider])
  async providers(@Arg('location', () => [Int]) location: [number, number]): Promise<Provider[]> {
    
    const providers: Provider[] = await ProviderModel.find();

    providers.map((provider) => {
      provider.distanceToProvider = getDistance(location, provider.locations[0].geocode);
    });

    return providers;
  }
```

**Reply by MichalLytek** [reply]
Field resolvers are for lazy resolving fields. It can be time-consuming fetching related data from database or cost-complex calculating some values. It's for all that you don't want to do in the query because it's unnecessary work when the field is not requested and also because field resolvers allows you to write the logic once and it works for every query and mutation that returns that type of data.

**Reply by cnps** [reply]
Thanks for the help! It cleared up a lot for me. Much appreciated!
